runtime/src/main/java/com/dylibso/chicory/runtime/ByteArrayMemory.java

```diff
@@ -62,145 +62,117 @@ public ByteArrayMemory(MemoryLimits limits, MemAllocStrategy allocStrategy) {
-        } catch (InterruptedException ie) {
-            Thread.currentThread().interrupt(); // Restore interrupt status
-            throw new ChicoryInterruptedException("Thread interrupted");
-        }
-        if (System.nanoTime() >= endTime) {
-            return 2; // timeout
-        } else {
-            return 0; // wake
-        }
+        return waitStates.computeIfAbsent(address, k -> new WaitState());
     }
 
-    private void endWaitOn(int address) {
-        AtomicInteger notifyCount = notifyInProgress.get(address);
-        if (notifyCount != null && notifyCount.decrementAndGet() == 0) {
-            notifyInProgress.remove(address);
-        }
-        AtomicInteger monitor = monitors.get(address);
-        if (monitor != null && monitor.decrementAndGet() == 0) {
-            monitors.remove(address);
-        }
-    }
-
-    // Wait until value at address != expected
-    @Override
-    public int waitOn(int address, int expected, long timeout) {
+    // Wait IF condition is true
+    private int waitOn(int address, BooleanSupplier condition, long timeout) {
         if (!shared()) {
             throw new ChicoryException("Attempt to wait on a non-shared memory, not supported.");
         }
-        AtomicInteger monitor = nextMonitor(address);
 
-        synchronized (monitor) {
+        long deadline = (timeout < 0) ? Long.MAX_VALUE : System.nanoTime() + timeout;
+
+        WaitState state = waitStates.computeIfAbsent(address, k -> new WaitState());
+
+        synchronized (state) {
+            // Check the condition while holding the lock
+            // This must be atomic with the decision to wait
+            if (!condition.getAsBoolean()) {
+                return 1; // not-equal
+            }
+
+            state.waiterCount++;
             try {
-                if (readInt(address) == expected) {
-                    return waitOnMonitor(
-                            address, (timeout < 0) ? Long.MAX_VALUE : timeout, monitor);
-                } else {
-                    return 1; // not-equal
+
+                while (state.pendingWakeups == 0) {
+                    long remaining = deadline - System.nanoTime();
+                    if (remaining <= 0) {
+                        return 2; // timeout
+                    }
+                    long millis = Math.max(remaining / 1_000_000L, 0);
+                    int nanos = Math.max((int) (remaining % 1_000_000L), 0);
+                    try {
+                        state.wait(millis, nanos);
+                    } catch (InterruptedException ie) {
+                        Thread.currentThread().interrupt();
+                        throw new ChicoryInterruptedException("Thread interrupted");
+                    }
                 }
+                // Consume one pending wakeup
+                state.pendingWakeups--;
+                return 0; // woken
             } finally {
-                endWaitOn(address);
+                state.waiterCount--;
```

Comment:
I feel like this is the right place to remove the entry from `waitStates` when both `pendingWakeups` and `waiterCount` are equals to 0.

Reply:
I tried that, but the problem is that other threads coming into wait have already read the waitState and are waiting to acquire it. It can be fixed by adding a `reclaimed` to the waitState that is set when it's removed from the map. If another thread acquires a waitState that happens to be reclaimed it can loop around and read from scratch.

I'm happy to add this to the PR if you want a look.